src/components/Authenicator/OTPDisplay.tsx

```diff
@@ -0,0 +1,180 @@
+                    alignItems: 'flex-end',
+                    minWidth: '120px',
+                    textAlign: 'right',
+                }}>
+                <p
+                    style={{
+                        fontWeight: 'bold',
+                        marginBottom: '0px',
+                        fontSize: '10px',
+                        marginTop: 'auto',
+                        textAlign: 'right',
+                    }}>
+                    next
+                </p>
+                <p
+                    style={{
+                        fontSize: '14px',
+                        fontWeight: 'bold',
+                        marginBottom: '0px',
+                        marginTop: 'auto',
+                        textAlign: 'right',
+                    }}>
+                    {nextCode}
+                </p>
+            </div>
+        </div>
+    );
+};
+
+function BadCodeInfo({ codeInfo, codeErr }) {
+    const [showRawData, setShowRawData] = useState(false);
+
+    return (
+        <div className="code-info">
+            <div>{codeInfo.title}</div>
+            <div>{codeErr}</div>
+            <div>
+                {showRawData ? (
+                    <div onClick={() => setShowRawData(false)}>
+                        {codeInfo.rawData ?? 'no raw data'}
+                    </div>
+                ) : (
+                    <div onClick={() => setShowRawData(true)}>Show rawData</div>
+                )}
+            </div>
+        </div>
+    );
+}
+
+interface OTPDisplayProps {
+    codeInfo: Code;
+}
+
+const OTPDisplay = (props: OTPDisplayProps) => {
+    const { codeInfo } = props;
+    const [code, setCode] = useState('');
+    const [nextcode, setNextCode] = useState('');
+    const [codeErr, setCodeErr] = useState('');
+
+    const generateCodes = () => {
+        try {
+            const currentTime = new Date().getTime();
+            if (codeInfo.type.toLowerCase() === 'totp') {
+                const totp = new TOTP({
+                    secret: codeInfo.secret,
+                    algorithm: codeInfo.algorithm ?? Code.defaultAlgo,
+                    period: codeInfo.period ?? Code.defaultPeriod,
+                    digits: codeInfo.digits ?? Code.defaultDigits,
+                });
+                setCode(totp.generate());
+                setNextCode(
+                    totp.generate({
+                        timestamp: currentTime + codeInfo.period * 1000,
+                    })
+                );
+            } else if (codeInfo.type.toLowerCase() === 'hotp') {
+                const hotp = new HOTP({
+                    secret: codeInfo.secret,
+                    counter: 0,
+                    algorithm: codeInfo.algorithm,
+                });
+                setCode(hotp.generate());
+                setNextCode(hotp.generate({ counter: 1 }));
+            }
+        } catch (err) {
+            setCodeErr(err.message);
+        }
+    };
```

Comment:
Didn't understood this part. I want to accomplish following behaviour:

1. Immediately call `generateCode` on the rendering of component
2. Call `generateCode` every 1 seconds so that after pre-defined period (30/60s etc), new code is generated.

Given that, can you suggest what changes shall I make?

Reply:
What I meant was that, 

we can have one function, `createCodeGeneration` which would just create and return the generator instance 
```
const createCodeGeneration = () => {
        if (codeInfo.type.toLowerCase() === 'totp') {
            return new TOTP({
                secret: codeInfo.secret,
                algorithm: codeInfo.algorithm ?? Code.defaultAlgo,
                period: codeInfo.period ?? Code.defaultPeriod,
                digits: codeInfo.digits ?? Code.defaultDigits,
            });
        } else if (codeInfo.type.toLowerCase() === 'hotp') {
            return new HOTP({
                secret: codeInfo.secret,
                counter: 0,
                algorithm: codeInfo.algorithm,
            });
        }
    };
```

and with that we can generate the currentCode by calling `generator.generate()` and for the next code we could have a helper function like `generateNextCodes`

```
 const generateNextCodes = (generator: TOTP | HOTP, type: Code['type']) => {
        if (type.toLowerCase() === 'totp') {
            const currentTime = new Date().getTime();
            return generator.generate({
                timestamp: currentTime + codeInfo.period * 1000,
            });
        } else if (codeInfo.type.toLowerCase() === 'hotp') {
            return generator.generate({ counter: 1 });
        }
    };
```

now the useEffect could be updated to and the two function `setupGenerator` and `generateNextCode` could be moved to util/service file 

```

const updateCodes = (generator: TOTP | HOTP) => {
        try {
            const code = generator.generate();
            const nextCode = generateNextCodes(generator, codeInfo.type);
            setCode(code);
            setNextCode(nextCode);
        } catch (err) {
            setCodeErr(err.message);
        }
    };

    useEffect(() => {
        let interval;
        try {
            const generator = createCodeGeneration();
            updateCodes(generator);
            interval = setInterval(() => {
                updateCodes(generator);
            }, generateCodeInterval);
        } catch (err) {
            setCodeErr(err.message);
        }
        return () => {
            clearInterval(interval);
        };
    }, []);

```

Reply:
This can be improved further, as we are unneccesarily trying to udpate Code every second. 

Which we can fix by updating the code something like this.

```

const setupCodeUpdateLoop = (generator: TOTP | HOTP) => {
        const microSecondsInPeriod = codeInfo.period * 1000;
        const timeToNextCode =
            microSecondsInPeriod -
            ((new Date().getTime() * 1000) % microSecondsInPeriod);

        let interval: NodeJS.Timeout;
        const timeout = setTimeout(() => {
            updateCodes(generator);
            interval = setInterval(() => {
                updateCodes(generator);
            }, microSecondsInPeriod);
        }, timeToNextCode);
        return {
            cleanUp: () => {
                clearTimeout(timeout);
                clearInterval(interval);
            },
        };
    };

    useEffect(() => {
        try {
            const generator = createCodeGenerator();
            updateCodes(generator);
            const { cleanUp } = setupCodeUpdateLoop(generator);
            return () => {
                cleanUp();
            };
        } catch (err) {
            setCodeErr(err.message);
        }
    }, []);
```

Reply:
(discussed offline, no action item needed)